man/file-hierarchy.xml

```diff
@@ -584,6 +589,18 @@
     directives of service units (see
     <citerefentry><refentrytitle>systemd.unit</refentrytitle><manvolnum>5</manvolnum></citerefentry>
     for details).</para>
+
+    <para><filename>/tmp/</filename>, <filename>/var/tmp/</filename> and <filename>/dev/shm/</filename>
+    should be mounted <option>nosuid</option> and <option>nodev</option>, which means that set-user-id mode
+    and character or block special devices are not interpreted on those file systems. But they should
+    <emphasis>not</emphasis> not be mounted <option>noexec</option>, becacuse various programs use those
+    directories for dynamically generated code or optimized code, and with that flag those use cases would
+    break. See the discussion of <option>nosuid</option>/<option>nodev</option>/<option>noexec</option>
```

Comment:
I don't think /tmp and /var/tmp are used for generated code but /dev/shm and possibly /run/user/$UID. I've been running with noexec /tmp for years without problems. I'd drop such advise.

Also "becacuse".

Reply:
The bug linked in the commit description is about Python ffi module failing when /tmp is noexec. So there clearly are cases and I think it's better to avoid the issue completely.

Reply:
In the bug report also /dev/shm was noexec but I don't think that can be assumed to be OK in the general case. In hardened setups like in the bug report there can be preferences to not use programs which generate code. But correctly written programs should use /dev/shm (not /tmp or /var/tmp) for generating executable files.

Reply:
Where did you get the idea that /dev/shm is meant for "generated code"? As far as I know, it exists to support glibc's shared memory implementation.

https://www.man7.org/linux/man-pages/man7/shm_overview.7.html

Reply:
The manual page doesn't say what other uses /dev/shm may have.

Reply:
The only other (ab)use I have seen is as a fallback for when other locations are locked with options that are too restrictive. It's a workaround, not an intended use.

Reply:
Perhaps more neutral advice would be that security considerations with having file systems which allow both writing and executing should be balanced with the needs to run certain classes of software which dynamically generate code.

Reply:
This PR is related to **libffi** RH bug https://bugzilla.redhat.com/show_bug.cgi?id=1875340
As of now, **libffi** tries to use /tmp, /var/tmp, /dev/shm and $HOME in that order to execute generated code.
Apparently Nessus security scanner is recommending setting noexec on /tmp, /var/tmp and /dev/shm.
STIG recommends having noexec on /dev/shm: RHEL-07-021024
There is also noexec recommended on /tmp, /var/tmp and /dev/shm for SLES15: https://www.cisecurity.org/benchmark/suse_linux/

Additionally, various kickstart delivered through scap-security-guide mount /tmp, /var/tmp and /dev/shm with noexec

Reply:
Many different people recommend different things. The question is whether something makes sense. We know that following the recommendation breaks things, and noexec on /dev/shm is probably not going to happen. And if you have no noexec on /dev/shm, the setting on /tmp doesn't matter much for security, because attackers can use /dev/shm instead of /tmp.

Would you like this better if I added "on general-purpose systems"?

Reply:
we previously had reports that some versions of Java do use this too to share JIT compiled versions of what they do. And I don't even think it's a bad idea to do that or disallow that. There's definitely value in sharing JIT compiled stuff between processes, and /dev/shm and friends are not inherently wrong for that. I mean, they need some place to store stuff that can be mapped executable, and I see no point in chasing developers away from these three places, there's nothing inherently better than these.

Hence i think this change makes sense, but we should emphasize that for the general case noexec is not OK, for specific cases it might be.